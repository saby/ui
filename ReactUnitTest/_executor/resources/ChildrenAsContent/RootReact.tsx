import { FC, ComponentType, ElementType } from 'react';

interface IRootReactComponentProps {
    WasabyComponentWithContent: ComponentType;
    Content?: ElementType<{ className?: string }>;
    contentClassName?: string;
}

const RootReactComponentVar: FC<IRootReactComponentProps> = function RootReactComponent({
    WasabyComponentWithContent,
    Content = 'div',
    contentClassName = 'content',
}: IRootReactComponentProps) {
    return (
        <div className="rootReactComponent">
            <WasabyComponentWithContent>
                <Content className={contentClassName}>Hello content</Content>
            </WasabyComponentWithContent>
        </div>
    );
};

export default RootReactComponentVar;
