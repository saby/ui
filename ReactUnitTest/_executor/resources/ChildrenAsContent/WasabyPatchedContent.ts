import { Control } from 'UICore/Base';
import template = require('wml!ReactUnitTest/_executor/resources/ChildrenAsContent/WasabyPatchedContent');

export default class WasabyPatchedContent extends Control {
    protected _template = template;
}
