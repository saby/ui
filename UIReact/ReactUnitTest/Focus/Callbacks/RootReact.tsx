import { FC, memo, forwardRef } from 'react';
import { TInternalProps } from 'UICore/Executor';
import { FocusArea, FocusRoot } from 'UICore/Focus';
import Input from 'ReactUnitTest/Focus/Input';
import ReactInput from 'ReactUnitTest/Focus/ReactInput';

const WasabyInput = Input;

interface IRootReact extends TInternalProps {
    onRootActivated: () => void;
    onRootDeactivated: () => void;
    onWasabyActivated: () => void;
    onReactDeactivated: () => void;
}

const RootReact: FC<IRootReact> = forwardRef((props: IRootReact, ref?) => {
    return (
        <FocusRoot
            as="div"
            {...props.attrs}
            ref={ref}
            onActivated={props.onRootActivated}
            onDeactivated={props.onRootDeactivated}
        >
            <FocusRoot as="div" onDeactivated={props.onReactDeactivated}>
                <ReactInput inputId="focusCallbacksReactId" />
            </FocusRoot>
            <div>
                <FocusArea onActivated={props.onWasabyActivated}>
                    <WasabyInput inputId="focusCallbacksWasabyId" />
                </FocusArea>
            </div>
        </FocusRoot>
    );
});

export default memo(RootReact);
