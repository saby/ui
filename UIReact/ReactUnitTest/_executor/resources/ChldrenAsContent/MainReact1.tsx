import InnerReact1 from './InnerReact1';
import Wasaby1 from './Wasaby1';
import * as React from 'react';

function MainReact1(props, ref) {
    const [i, seti] = React.useState(0);
    function clickHandler() {
        seti(i + 1);
    }
    return (
        <Wasaby1 value={i}>
            <div>
                <button onClick={clickHandler}>+1</button>
                <InnerReact1 value={i} fref={ref}></InnerReact1>
            </div>
        </Wasaby1>
    );
}
export default React.memo(React.forwardRef(MainReact1));
