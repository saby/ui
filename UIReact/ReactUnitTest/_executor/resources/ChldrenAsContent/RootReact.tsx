import { FC, ComponentType, ElementType } from 'react';

interface IRootReactComponentProps {
    WasabyComponentWithContent: ComponentType<unknown>;
    Content?: ElementType<{ className: string }>;
}

const RootReactComponentVar: FC<IRootReactComponentProps> = function RootReactComponent({
    WasabyComponentWithContent,
    Content = 'div',
}: IRootReactComponentProps) {
    return (
        <div className="rootReactComponent">
            <WasabyComponentWithContent>
                <Content className="content">Hello content</Content>
            </WasabyComponentWithContent>
        </div>
    );
};

export default RootReactComponentVar;
