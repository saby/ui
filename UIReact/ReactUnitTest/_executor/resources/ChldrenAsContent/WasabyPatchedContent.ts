import { Control } from 'UICore/Base';
import template = require('wml!ReactUnitTest/_executor/resources/ChldrenAsContent/WasabyPatchedContent');

export default class WasabyPatchedContent extends Control {
    protected _template = template;
}
