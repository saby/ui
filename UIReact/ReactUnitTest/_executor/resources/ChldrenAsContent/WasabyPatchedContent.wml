<div class="wasabySimpleContent">
   <ws:partial template="{{ _options.content }}" attr:class="patchedClass" anotherOption="value" />
</div>
