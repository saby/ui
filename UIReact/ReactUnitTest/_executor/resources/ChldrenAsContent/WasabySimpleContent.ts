import { Control } from 'UICore/Base';
import template = require('wml!ReactUnitTest/_executor/resources/ChldrenAsContent/WasabySimpleContent');

export default class WasabySimpleContent extends Control {
    protected _template = template;
}
