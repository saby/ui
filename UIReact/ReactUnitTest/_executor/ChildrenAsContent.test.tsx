/**
 * @jest-environment jsdom
 */
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';

import RootReact from './resources/ChldrenAsContent/RootReact';
import WasabySimpleContent from './resources/ChldrenAsContent/WasabySimpleContent';
import WasabyPatchedContent from './resources/ChldrenAsContent/WasabyPatchedContent';
import WasabySimpleContentTemplate = require('wml!ReactUnitTest/_executor/resources/ChldrenAsContent/WasabySimpleContent');
import WasabyPatchedContentTemplate = require('wml!ReactUnitTest/_executor/resources/ChldrenAsContent/WasabyPatchedContent');

import MainReact1 from './resources/ChldrenAsContent/MainReact1';
import MainReact2 from './resources/ChldrenAsContent/MainReact2';

async function wait() {
    await act(async () => {
        jest.runOnlyPendingTimers();
    });
}

describe('Children в tsx преобразовывается в content wml', () => {
    let container: HTMLDivElement;
    beforeEach(() => {
        container = document.createElement('div', {});
        document.body.appendChild(container);
    });
    afterEach(() => {
        unmountComponentAtNode(container);
        container.remove();
        container = null;
    });

    it('контрол не патчит content', () => {
        act(() => {
            render(
                <RootReact WasabyComponentWithContent={WasabySimpleContent} />,
                container
            );
        });
        expect(container).toMatchSnapshot();
    });

    it('контрол патчит content', () => {
        act(() => {
            render(
                <RootReact WasabyComponentWithContent={WasabyPatchedContent} />,
                container
            );
        });
        expect(container).toMatchSnapshot();
    });

    it('шаблон не патчит content', () => {
        act(() => {
            render(
                <RootReact
                    WasabyComponentWithContent={WasabySimpleContentTemplate}
                />,
                container
            );
        });
        expect(container).toMatchSnapshot();
    });

    it('шаблон патчит content', () => {
        act(() => {
            render(
                <RootReact
                    WasabyComponentWithContent={WasabyPatchedContentTemplate}
                />,
                container
            );
        });
        expect(container).toMatchSnapshot();
    });

    it('не происходит лишних перерисовок 1 (перерисовка по клику происходит и счетчик выводится)', async () => {
        act(() => {
            render(<MainReact1 />, container);
        });
        await wait();

        const button = container.getElementsByTagName('button')[0];

        expect(container).toMatchSnapshot();
        button.click();
        expect(container).toMatchSnapshot();
    });
    it('не происходит лишних перерисовок 2 (перерисовка по клику не происходит)', async () => {
        act(() => {
            render(<MainReact2 />, container);
        });
        await wait();

        const button = container.getElementsByTagName('button')[0];

        expect(container).toMatchSnapshot();
        button.click();
        expect(container).toMatchSnapshot();
    });
});
