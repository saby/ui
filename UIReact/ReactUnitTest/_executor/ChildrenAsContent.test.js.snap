// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Children в tsx преобразовывается в content wml tsx вставляет Wasaby в Wasaby: Мержатся все 4 класса, объявленные на всех уровнях 1`] = `
<div>
  <div>
    <div
      class="inner-class react-inner patched-wasaby-class react-outer"
      data-text="data-text from OuterWasaby.wml"
      tabindex="0"
    >
      Text from WasabyInWasaby.tsx
    </div>
  </div>
</div>
`;

exports[`Children в tsx преобразовывается в content wml Объединение ref, в Wasaby вставляется HTMLElement: Проверка вёрстки 1`] = `
<div>
  <div
    class="rootReactForwardingRef"
  >
    <div
      class="wasabyContentInRoot contentWithRef"
    >
      Hello content
    </div>
  </div>
</div>
`;

exports[`Children в tsx преобразовывается в content wml Объединение ref, в Wasaby вставляется Wasaby: Проверка вёрстки 1`] = `
<div>
  <div
    class="rootReactForwardingRef"
  >
    <div
      class="inner-class contentWithRef wasabyContentInRoot"
      tabindex="0"
    >
      Wasaby
    </div>
  </div>
</div>
`;

exports[`Children в tsx преобразовывается в content wml Происходит склейка className когда контент это реакт компонент: должны склеиться классы patchedClass и content 1`] = `
<div>
  <div
    class="rootReactComponent"
  >
    <div
      class="wasabySimpleContent"
      tabindex="0"
    >
      <div
        class="patchedClass content"
      >
        Hello content
      </div>
    </div>
  </div>
</div>
`;

exports[`Children в tsx преобразовывается в content wml контрол не патчит content 1`] = `
<div>
  <div
    class="rootReactComponent"
  >
    <div
      class="wasabySimpleContent"
      tabindex="0"
    >
      <div
        class="content"
      >
        Hello content
      </div>
    </div>
  </div>
</div>
`;

exports[`Children в tsx преобразовывается в content wml контрол патчит content 1`] = `
<div>
  <div
    class="rootReactComponent"
  >
    <div
      class="wasabySimpleContent"
      tabindex="0"
    >
      <div
        class="patchedClass content"
      >
        Hello content
      </div>
    </div>
  </div>
</div>
`;

exports[`Children в tsx преобразовывается в content wml контрол прокидывает content дальше 1`] = `
<div>
  <div
    class="rootReactComponent"
  >
    <div
      class="wasabySimpleContent wasaby-proxy-children"
      tabindex="0"
    >
      <div
        class="content"
      >
        Hello content
      </div>
    </div>
  </div>
</div>
`;

exports[`Children в tsx преобразовывается в content wml не происходит лишних перерисовок 1 (перерисовка по клику происходит и счетчик выводится) 1`] = `
<div>
  <div>
    <button>
      +1
    </button>
    <div>
      0
    </div>
  </div>
</div>
`;

exports[`Children в tsx преобразовывается в content wml не происходит лишних перерисовок 1 (перерисовка по клику происходит и счетчик выводится) 2`] = `
<div>
  <div>
    <button>
      +1
    </button>
    <div>
      1
    </div>
  </div>
</div>
`;

exports[`Children в tsx преобразовывается в content wml не происходит лишних перерисовок 2 (перерисовка по клику не происходит) 1`] = `
<div>
  <div>
    <button>
      +1
    </button>
    <div>
      0
    </div>
  </div>
</div>
`;

exports[`Children в tsx преобразовывается в content wml не происходит лишних перерисовок 2 (перерисовка по клику не происходит) 2`] = `
<div>
  <div>
    <button>
      +1
    </button>
    <div>
      0
    </div>
  </div>
</div>
`;

exports[`Children в tsx преобразовывается в content wml пропсы из логического родителя в приоритете перед опциями контента: anotherOption не перетирается в wml 1`] = `
<div>
  <div
    class="wasabySimpleContent"
    tabindex="0"
  >
    <div
      class="patchedClass"
    >
      anotherOption: tsx value
    </div>
  </div>
</div>
`;

exports[`Children в tsx преобразовывается в content wml шаблон не патчит content 1`] = `
<div>
  <div
    class="rootReactComponent"
  >
    <div
      class="wasabySimpleContent"
      tabindex="0"
    >
      <div
        class="content"
      >
        Hello content
      </div>
    </div>
  </div>
</div>
`;

exports[`Children в tsx преобразовывается в content wml шаблон патчит content 1`] = `
<div>
  <div
    class="rootReactComponent"
  >
    <div
      class="wasabySimpleContent"
      tabindex="0"
    >
      <div
        class="patchedClass content"
      >
        Hello content
      </div>
    </div>
  </div>
</div>
`;
